### ISSUE TITLE
Training scripts / code?

### ISSUE THREAD

**Issue by jleem-iso** [brennanaba/ABlooper#10]
Hi Brennan, have you got plans to release how you guys trained the model? 

It'd help reproducibility and also help users train models on bespoke sets of structures. There were other details in the paper (e.g. batching) which weren't discussed in the paper but could be of importance for the model's performance.

**Comment by brennanaba** [comment]
Hi Jin,

I don't intend to make the training code publicly available. This is partly because it depends on some in-house code (like the Sabdab API). But also, it would take me some time to bring the code up to a sharable standard and I don't think it would benefit most users. 

I used a batch size of 4 for the final model, but I don't think that had a big effect on the results. Sorry for forgetting to include that in the paper. 

Happy to have a chat if you have any questions though!! 

All the best,

Brennan

**Comment by jleem-iso** [comment]
Thanks Brennan. There's still details that I think the paper doesn't really go through, including:

- LR scheduling (was it flat? was it decayed, etc)
- I think you mentioned 100? 200? Epochs in the SI, but where along do you find you need to terminate early?
- did you account for obsolete structures? SAbDab has both current and obsolete structures from my memory and that could duplicate essentially the same example twice (of course, having sequence identical but conformationally unique examples is a good thing)
- did you use all chains / units within an ASU?
- when you say early stopping that's presumably at the point with the lowest validation set loss; but was it the lowest joint loss or lowest of one of the two losses you used?
- when there are 5 independent networks, did you just train them with different seeds? Did you find that their loss trajectories were substantially different?

If you're not happy with your own trainer, I can have a crack at doing something using no in-house code (i.e. just give it a bunch of PDB files) and have a train/eval loop time permitting...

I disagree it wouldn't be valuable, because there have been plenty of cases (e.g. AlphaFold) where people have essentially recreated the training procedure.

**Comment by brennanaba** [comment]
Hi Jin,

- Flat LR. 
- At what point it stopped is run-dependent. The validation loss was checked after each epoch and if it did not improve after 100 epochs it was stopped.
- I don't know about obsolete structures, but I am pretty sure I did not use them during training. 
- I don't know what an ASU is.
- It was the loss that was being used at that stage of training.
- They were initialised with different seeds. E(n)-EGNN are quite unstable, so the loss trajectories were often different and in some runs never converged.

All the best,

Brennan

**Comment by jleem-iso** [comment]
Ah thanks. This is helpful. I was curious about the decision to leave out
training code more than anything - again given the prevalence of training
scripts elsewhere I thought it's a no-brainer decision.

I meant asymmetric units, as in cases like the structure 5ggs, which has
multiple chains, did you use all of them or only a couple, etc.

I think obsolete structures are hard to come by, but it can be checked via
the PDB API. Great news if they weren't used, of course!

Would you say the stability was one decision behind using EGNNs over
something like the SE3 transformer? It's really interesting to see the
growth and diversity of different strategies to handle equivariance (e.g.
Equiformer, or using IPA attention, etc).

Sorry one more question - what was behind the decision in leaving out
backbone oxygen atoms?

Best wishes

Jin

**Comment by brennanaba** [comment]
Hi Jin,

Sorry for the late reply, I have been quite busy the last few days.

If a PDB file in the training set had multiple Fvs, then I used all of them.

I found the SE3 transformer to be more stable than EGNNs during training.

The oxygen position can be predicted pretty well from the main chain atoms.

All the best,

Brennan
